File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer[]}
def preorder_traversal(root)
    if root == nil
        return []
    end
    s = []
    s.push(root.val)
    preorder_traversal(root.left).each do |x|
        s.push(x)
    end
    preorder_traversal(root.right).each do |x|
        s.push(x)
    end
    return s
end

File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var preorderTraversal = function(root)
{
    var r = [];
    if(root == null)
    {
        return r;
    }
    traverse(root, r);
    return r;
};

var traverse = function(rt, r)
{
    r.push(rt.val);
    if(rt.left != null)
    {
        traverse(rt.left, r);
    }
    if(rt.right != null)
    {
        traverse(rt.right, r);
    }
};